pkg/election/election.go

```diff
@@ -0,0 +1,251 @@
+func (e *Election) ErrorNotify() <-chan error {
+	return e.ech
+}
+
+// Close closes the election instance and release the resources.
+func (e *Election) Close() {
+	e.l.Info("election is closing")
+	if !e.closed.CompareAndSwap(0, 1) {
+		e.l.Info("election was already closed")
+		return
+	}
+
+	e.cancel()
+	e.bgWg.Wait()
+	e.l.Info("election is closed")
+}
+
+func (e *Election) campaignLoop(ctx context.Context) {
+	var (
+		session *concurrency.Session
+		err     error
+	)
+
+	closeSession := func(se *concurrency.Session) {
+		err2 := se.Close() // only log this error
+		if err2 != nil {
+			e.l.Error("fail to close etcd session", zap.Int64("lease", int64(se.Lease())), zap.Error(err2))
+		}
+	}
+
+	defer func() {
+		if session != nil {
+			closeSession(session) // close the latest session.
+		}
+		if err != nil && errors.Cause(err) != ctx.Err() { // only send non-ctx.Err() error
+			e.l.Error("", zap.Error(err))
+			e.ech <- err
+		}
+	}()
+
+	// create the initial session.
+	session, err = newSession(e.cli, e.sessionTTL)
+	if err != nil {
+		err = terror.ErrElectionCampaignFail.Delegate(err, "create the initial session")
+		return
+	}
+
+	for {
+		// check context canceled/timeout
+		select {
+		case <-session.Done():
+			e.l.Info("etcd session is done, will try to create a new one", zap.Int64("old lease", int64(session.Lease())))
+			closeSession(session)
+			session, err = newSession(e.cli, e.sessionTTL)
+			if err != nil {
+				err = terror.ErrElectionCampaignFail.Delegate(err, "create a new session")
+				return
+			}
+		case <-ctx.Done():
+			e.l.Info("break campaign loop, context is done", zap.Error(ctx.Err()))
+			return
+		default:
+		}
+
+		// try to campaign
+		elec := concurrency.NewElection(session, e.key)
+		err = elec.Campaign(ctx, e.id)
+		if err != nil {
+			// err may be ctx.Err(), but this can be handled in `case <-ctx.Done()`
+			e.l.Warn("fail to campaign", zap.Error(err))
+			continue
+		}
+
+		// compare with the current leader
+		leaderKey, leaderID, err := getLeaderInfo(ctx, elec)
+		if err != nil {
+			// err may be ctx.Err(), but this can be handled in `case <-ctx.Done()`
+			e.l.Warn("fail to get leader ID", zap.Error(err))
+			continue
+		}
+		if leaderID != e.id {
+			e.l.Info("current member is not the leader", zap.String("current member", e.id), zap.String("leader", leaderID))
+			continue
+		}
+
+		e.toBeLeader() // become the leader now
+		e.watchLeader(ctx, session, leaderKey)
+		e.retireLeader() // need to re-campaign
+	}
+}
+
+func (e *Election) toBeLeader() {
+	e.isLeader.Set(true)
+	select {
+	case e.leaderCh <- true:
+	default:
+	}
+	e.l.Info("become the leader", zap.String("member", e.id))
+}
+
+func (e *Election) retireLeader() {
+	e.isLeader.Set(false)
+	select {
+	case e.leaderCh <- false:
+	default:
+	}
+	e.l.Info("retire from the leader", zap.String("member", e.id))
+}
+
+func (e *Election) watchLeader(ctx context.Context, session *concurrency.Session, key string) {
```

Comment:
the key can only be e.key? if so don't need put it in arguments.

Reply:
`e.key` is the prefix of this `key`, like `e.key == /dm-master/leader`, but `key == /dm-master/leader/105a6e8c8aee0806` (with the `lease` as the suffix), and this key can change for every campaign loop.

Reply:
ok